### ISSUE TITLE
Question about position encoding

### ISSUE THREAD

**Issue by codejoker-c** [manycore-research/SpatialLM#78]
Hello authors,

Thank you for your excellent work and for sharing your code!

I have a question regarding the positional encoding strategy used in the model. From the code, I noticed that after obtaining the Sonata encoder features, the normalized grid_coord is used to generate the Fourier positional encoding. 

```
        point = self.enc(point)
        context = point["sparse_conv_feat"].features

        if self.enable_fourier_encode:
            coords = point["grid_coord"]
            coords_normalised = coords / (self.reduced_grid_size - 1)
            encoded_coords = fourier_encode_vector(coords_normalised)

            context = torch.cat([context, encoded_coords], dim=-1)
            context = self.input_proj(context)
```

However, I'm a bit confused about this choice. Since the input point cloud undergoes positive shifting and other preprocessing steps to compute grid_coord, it seems that the resulting coordinates may not be fully aligned with the original point cloud. This leads to the misalignment between the predict bbox and the input point cloud bbox. Could you please clarify the rationale behind this design choice? Or let me know if there is something I might have misunderstood?

Looking forward to your insights. Thanks again for your great work!

**Comment by SamMaoYS** [comment]
Hi, thank you for your question. The normalized grid coordinates represent the voxel indices in the deepest layer of the encoder. For example, if the original point cloud `grid_coord` along the x-axis is 128, after passing through four downsampling layers in the encoder (each reducing the size by half), the final x-axis index becomes 128 / (2^4) = 8. We followed the same positional encoding strategy as [SceneScript ](https://github.com/facebookresearch/scenescript/blob/571dbe20a19822517722ec3a99a054870c3362e1/src/networks/encoder.py#L276). 

In our experiments, we tested using only the Sonata encoded features, without including these relative positional embeddings, and observed noticeable misalignments between the predicted bounding boxes and the input point cloud. Our interpretation is that U-Net based method mainly leverage skip connections to pass XYZ from the input point cloud to the Decoder blocks, to remap features back to their accurate XYZ positions. The encoder primarily extracts semantic features, while positional information is largely restored in the decoder via these skip connections using the ground-truth point coordinates. It generally works well for segmentation tasks, or methods with a decoder.

Therefore, adding positional embeddings in the last layer helps the model predict more accurate position values in our cases. We have not yet explored more advanced positional embedding strategies, such as learnable Fourier embeddings or Rotary Positional Embeddings (RoPE), which are interesting directions for future work.

**Comment by SamMaoYS** [comment]
Sure, for the first question:
- Yes, the ground truth annotations are shifted, normalized, and discretized to align with the grid sampling or any other augmentations during training. You can check the [training data augmentation](https://github.com/manycore-research/SpatialLM/blob/22a08ce9d12ef72e409f7d812790e004dd1691e7/spatiallm/tuner/data/mm_plugin.py#L225) code here .

As for the second question:
- As the `grid_coords` are not the actual XYZ values of the points, but the indices of the voxels. And the normalization here is not a min-max normalization, but normalize with a global max number of bins. This global scale information should be easy to recover during the training, as well as the mapping between XYZ values and the voxel indices.
